fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobRunner.java

```diff
@@ -0,0 +1,308 @@
+                db.writeUnlock();
+            }
+        } catch (Throwable t) {
+            LOG.warn("create temp partitions failed", t);
+            throw t;
+        }
+    }
+
+    private void gc() {
+        LOG.info("start to garbage collect");
+        db.writeLock();
+        try {
+            if (job.getNewPartitionNames() != null) {
+                for (String partitionName : job.getNewPartitionNames()) {
+                    LOG.info("drop partition:{}", partitionName);
+
+                    Partition partition = targetTable.getPartition(partitionName, true);
+                    if (partition != null) {
+                        targetTable.dropTempPartition(partitionName, true);
+                    } else {
+                        LOG.warn("partition is null for name:{}", partitionName);
+                    }
+                }
+            }
+        } catch (Exception e) {
+            LOG.warn("exception when gc insert overwrite job.", e);
+        } finally {
+            db.writeUnlock();
+        }
+    }
+
+    private void doCommit() {
+        db.writeLock();
+        try {
+            if (targetTable.getPartitionInfo().getType() == PartitionType.RANGE) {
+                targetTable.replaceTempPartitions(job.getSourcePartitionNames(), job.getNewPartitionNames(), true, false);
+            } else {
+                targetTable.replacePartition(job.getSourcePartitionNames().get(0), job.getNewPartitionNames().get(0));
+            }
+        } catch (Exception e) {
+            LOG.warn("replace partitions failed when insert overwrite into dbId:{}, tableId:{}," +
+                            " sourcePartitionNames:{}, newPartitionNames:{}", job.getTargetDbId(), job.getTargetTableId(),
+                    job.getSourcePartitionNames().stream().collect(Collectors.joining(",")),
+                    job.getNewPartitionNames().stream().collect(Collectors.joining(",")), e);
+            throw new RuntimeException("replace partitions failed", e);
+        } finally {
+            db.writeUnlock();
+        }
+    }
+
+    private void prepareInsert() {
+        Preconditions.checkState(job.getJobState() == InsertOverwriteJobState.OVERWRITE_RUNNING);
+        Preconditions.checkState(insertStmt != null);
+        try {
+            db.readLock();
+            try {
+                List<Long> newPartitionIds = job.getNewPartitionNames().stream()
+                        .map(partitionName -> targetTable.getPartition(partitionName, true).getId())
+                        .collect(Collectors.toList());
+                PartitionNames partitionNames = new PartitionNames(true, job.getNewPartitionNames());
+                insertStmt.setOriginalTargetPartitionIds(insertStmt.getTargetPartitionIds());
+                insertStmt.setTargetPartitionNames(partitionNames);
+                insertStmt.setTargetPartitionIds(newPartitionIds);
```

Comment:
why need db.lock here?

Reply:
only one insert for all related partition?

Reply:
I think targetTable.getPartition should be called in read lock